### PR TITLE
[ONNX][TORCH-MLIR] LayerNorm

### PR DESCRIPTION
**Pull request by afalkenberg1**
Layer Normalization using the torch.aten.native_layer_norm

### CODE REVIEW

**Comment by afalkenberg1** [comment]
Hello 
@stellaraccident @vivekkhandelwal1 @rsuderman
It seems that this is finally working. I am using the native_layer_norm op which seems to do the correct thing. The layer_norm seems to have some limitations. 
Thanks

File: include/torch-mlir/Conversion/TorchOnnxToTorch/Patterns.h [hunk 1]
```diff
@@ -95,6 +95,16 @@ struct OpBinder {
...
+  ParseResult tensorResultTypeAtIndex(Torch::ValueTensorType &typeIdx, int64_t idx) {
+    if (idx >= op->getNumResults())
+      return failure();
+    auto t = toValidTensorType(op->getResult(idx).getType());
+    if (!t)
+      return failure();
+    typeIdx = t;
+    return success();
+  }
+
```
Comment (vivekkhandelwal1) [inline include/torch-mlir/Conversion/TorchOnnxToTorch/Patterns.h:107]: This code is already added here: https://github.com/llvm/torch-mlir/commit/80bd093d56c5f7b36c6852fdff05afac0d7c3a00

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp [hunk 2]
```diff
@@ -410,6 +410,50 @@ void mlir::torch::onnx_c::populateDefaultDomainGtoP(
...
+                      binder.tensorOperandAtIndex(B, 2) ||
+                      binder.tensorResultTypeAtIndex(Y_type, 0)) 
+                    return failure();
+                  if (binder.tensorResultTypeAtIndex(Mean_type, 1) ||
+                      binder.tensorResultTypeAtIndex(InvStdDev_type, 2))
+                    return failure();
+                  if (binder.s64IntegerAttr(axis, "axis", -1) ||
+                      binder.f32FloatAttr(epsilon, "epsilon", 0.00001) ||
+                      binder.s64IntegerAttr(stash_type, "stash_type", 1)) 
+                    return failure();
```
Comment (vivekkhandelwal1) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:435]: All these conditions can be merged.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp [hunk 3]
```diff
@@ -410,6 +410,50 @@ void mlir::torch::onnx_c::populateDefaultDomainGtoP(
...
+                      binder.tensorResultTypeAtIndex(InvStdDev_type, 2))
+                    return failure();
+                  if (binder.s64IntegerAttr(axis, "axis", -1) ||
+                      binder.f32FloatAttr(epsilon, "epsilon", 0.00001) ||
+                      binder.s64IntegerAttr(stash_type, "stash_type", 1)) 
+                    return failure(); 
+                  Value constEpsilon = rewriter.create<Torch::ConstantFloatOp>(
+                    binder.getLoc(), rewriter.getType<Torch::FloatType>(),
+                    rewriter.getF64FloatAttr(epsilon));
+                  unsigned rank = 1;
```
Comment (vivekkhandelwal1) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:439]: Is this based on some assumption?

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp [hunk 4]
```diff
@@ -410,6 +410,50 @@ void mlir::torch::onnx_c::populateDefaultDomainGtoP(
...
+                  if(axis < 0)
+                    axis = axis + rank;
+                  for (int64_t n = axis; n < rank ; n++) {
+                    normalized.push_back(rewriter.create<Torch::ConstantIntOp>(
+                    binder.getLoc(), rewriter.getI64IntegerAttr(n)));
+                  }
+                  Value normalized_shape = rewriter.create<Torch::PrimListConstructOp>(
+                    binder.getLoc(),
+                    Torch::ListType::get(Torch::IntType::get(binder.op->getContext())),
+                    normalized);
```
Comment (vivekkhandelwal1) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:452]: This logic for computing the `normalizedShape` seems incorrect.

Comment (vivekkhandelwal1) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:452]: Can you please tell us how did you infer this?

**Review by vivekkhandelwal1 — CHANGES_REQUESTED** [review]

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp (hunk 2 above)
Comment (afalkenberg1) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:435]: ok will do ; yes I had them separated to not loose the overview but now I can merge them.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp (hunk 4 above)
Comment (afalkenberg1) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:452]: See links above. The normlaizedShape is derived from axix. In onnx  you have axis and rank whereas in torch there is normalized shape. Please see the description in 
https://onnx.ai/onnx/operators/onnx__LayerNormalization.html

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp (hunk 3 above)
Comment (afalkenberg1) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:439]: I tried to follow the following: 
https://pytorch.org/docs/stable/generated/torch.nn.LayerNorm.html#torch.nn.LayerNorm
and 
https://onnx.ai/onnx/operators/onnx__LayerNormalization.html
There is a default mentioned for eps.

**Comment by afalkenberg1** [comment]
Here is some information
https://onnx.ai/onnx/operators/onnx__LayerNormalization.html

and 
https://pytorch.org/docs/stable/generated/torch.nn.LayerNorm.html

File: include/torch-mlir/Conversion/TorchOnnxToTorch/Patterns.h (hunk 1 above)
Comment (vivekkhandelwal1) [reply include/torch-mlir/Conversion/TorchOnnxToTorch/Patterns.h]: You can create different branches for different PRs.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp (hunk 3 above)
Comment (vivekkhandelwal1) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:437]: Sorry for the ambiguity. I wanted to ask about the rank being taken 1. How are we inferring that?

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp (hunk 4 above)
Comment (vivekkhandelwal1) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:450]: Although the downstream pipeline does not make use of normalized shape values except the list size still based on this https://onnx.ai/onnx/operators/onnx__LayerNormalization.html#:~:text=If%20X%E2%80%99s%20shape%20is%20%5Bd%5B0%5D%2C%20...%2C%20d%5Baxis%2D1%5D%2C%20d%5Baxis%5D%2C%20...%2C%20d%5Brank%2D1%5D%5D%2C%20the%20shape%20of%20Mean%20and%20InvStdDev%20is%20%5Bd%5B0%5D%2C%20...%2C%20d%5Baxis%2D1%5D%2C%201%2C%20...%2C%201%5D.%20Y%20and%20X%20have%20the%20same%20shape.

The normalized shape list should contain all 1s.

**Review by afalkenberg1 — COMMENTED** [review]
I hope this resolves the requested changes

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp [hunk 5]
```diff
@@ -410,6 +410,48 @@ void mlir::torch::onnx_c::populateDefaultDomainGtoP(
...
+                      binder.s64IntegerAttr(stash_type, "stash_type", 1)) 
+                    return failure(); 
+                  Value constEpsilon = rewriter.create<Torch::ConstantFloatOp>(
+                    binder.getLoc(), rewriter.getType<Torch::FloatType>(),
+                    rewriter.getF64FloatAttr(epsilon));
+                  unsigned rank = 1; 
+                  if(std::optional<unsigned> maybeRank = Torch::getTensorRank(X))
+                    rank = *maybeRank;
+                  SmallVector<Value> normalized;
+                  if(axis < 0)
```
Comment (AmosLewis) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp:441]: `axis = Torch::toPositiveDim(axis, inputRank);`

**Comment by AmosLewis** [comment]
Please update the onnx op tracker https://github.com/nod-ai/SHARK-Turbine/issues/325 and mark the op is taken when you start to implement a new op. Otherwise, there might be redundant work from others.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp (hunk 5 above)
Comment (afalkenberg1) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp]: ok thanks for the input ; changed it accordingly.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp (hunk 2 above)
Comment (afalkenberg1) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp]: merged them !

File: include/torch-mlir/Conversion/TorchOnnxToTorch/Patterns.h (hunk 1 above)
Comment (afalkenberg1) [reply include/torch-mlir/Conversion/TorchOnnxToTorch/Patterns.h]: Yes should I start this one in a new branch ? Which would lead to a new PR. I understand that it is a bit cleaner that way, but I created the first PR as an interim step and the LayerNorm depended on that so by the time the first PR was accepted and merged I already moved on with the LayerNorm.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp (hunk 4 above)
Comment (afalkenberg1) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainGtoP.cpp]: I resolved this issue please check the code. Yet I am understanding it in the way that normalized_shape contains dimensions from the input dimension. So if input is X[4,10,20,30] and axis is provided as maybe 2 then normalized_shape should be something like [20,30].

**Review by vivekkhandelwal1 — APPROVED** [review]
